FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift

```diff
@@ -0,0 +1,266 @@
+//
+//  DeviceBindingAuthenticators.swift
+//  FRAuth
+//
+//  Copyright (c) 2022 ForgeRock. All rights reserved.
+//
+//  This software may be modified and distributed under the terms
+//  of the MIT license. See the LICENSE file for details.
+//
+
+import Security
+import Foundation
+import FRCore
+import JOSESwift
+
+
+/// Protiocol  to override keypair generation, authentication, signing and access control
+public protocol DeviceAuthenticator {
+    
+    /// Generate  public and private key pair
+    func generateKeys() throws -> KeyPair
+    
+    /// Display authentication  prompt for authentication type if needed
+    /// - Parameter timeout: Timeout for teh authentication prompt
+    /// - Parameter completion: Completion block for Device binding result callback
+    func authenticate(timeout: Int, completion: @escaping DeviceBindingResultCallback)
+    
+    /// Sign the challenge sent from the server and generate signed JWT
+    /// - Parameter keyPair: Public and private key pair
+    /// - Parameter kid: Generated key id
+    /// - Parameter userId: user Id received from server
+    /// - Parameter challenge: challenge received from server
+    /// - Returns: compact serialized jws
+    func sign(keyPair: KeyPair, kid: String, userId: String, challenge: String) throws -> String
+    
+    /// Check if  authentication is supported
+    func isSupported() -> Bool
+    
+    /// Access Control for the authetication type
+    func accessControl() -> SecAccessControl?
+}
+
+
+extension DeviceAuthenticator {
+    
+    /// Default implemention
+    /// Sign the challenge sent from the server and generate signed JWT
+    /// - Parameter keyPair: Public and private key pair
+    /// - Parameter kid: Generated key id
+    /// - Parameter userId: user Id received from server
+    /// - Parameter challenge: challenge received from server
+    /// - Returns: compact serialized jws
+    func sign(keyPair: KeyPair, kid: String, userId: String, challenge: String) throws -> String {
+        
+        let jwk = try RSAPublicKey(publicKey: keyPair.publicKey, additionalParameters: ["use": "sig", "alg": "RS512"])
+        let jwkWithKeyId = try  jwk.withThumbprintAsKeyId()
+        
+        //create header
+        var header = JWSHeader(algorithm: .RS512)
+        header.kid = kid
+        header.typ = "JWS"
+        header.jwkTyped = jwkWithKeyId
+        
+        //create payload
+        let params: [String: Any] = ["sub": userId, "challenge": challenge, "exp": (Int(Date().addingTimeInterval(30).timeIntervalSince1970))]
```

Comment:
The `exp` value should be set with the value from the `timeout` attribute coming with the callback. 
In android, the following logic is in place:

>   /**
>      * Get Expiration date for the signed token, claim "exp" will be set to the JWS.
>      *
>      * @return The expiration date
>      */
>     protected open fun getExpiration(): Date {
>         val date = Calendar.getInstance();
>         date.add(Calendar.SECOND, timeout ?: 60)
>         return date.time;
>     }